{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c6343fa7-054b-4e88-8cd2-5b24e840433b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import anndata as an\n",
    "import pyranges as pr\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a856988b-5c71-44fe-9e31-f99674fec989",
   "metadata": {},
   "source": [
    "# Function Definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fb4804d2-0c50-49a5-bc67-45213ef2d04a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def feature_barcode_map(path, column_rename_map):\n",
    "    \"\"\"\n",
    "    Load and process a feature barcode map (e.g., fbc_map.csv) from the ONT_10x_feature_barcodes pipeline output.\n",
    "\n",
    "    This function:\n",
    "      - Reads a CSV file containing per-read feature barcode predictions.\n",
    "      - Parses and extracts condition names from the 'pred' column (e.g., 'chunk_0_MYOD' → 'MYOD').\n",
    "      - Aggregates feature barcode counts per cell and per condition.\n",
    "      - Identifies the most frequent (dominant) condition assigned to each cell.\n",
    "      - Computes the percentage of feature barcode reads supporting the dominant condition.\n",
    "      - Optionally renames columns based on a user-provided mapping.\n",
    "\n",
    "    Args:\n",
    "        path (str): \n",
    "            Path to the feature barcode map CSV file. \n",
    "            The file must contain at least 'cell_barcode' and 'pred' columns.\n",
    "        column_rename_map (dict): \n",
    "            Dictionary to rename condition columns (before or after pivot). \n",
    "            For example: {\"MYOD\": \"MYOD-fb_counts\", \"PRRX1\": \"PRRX1-fb_counts\"}\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: \n",
    "            A dataframe indexed by `cell_barcode` with:\n",
    "                - One column per condition (read counts),\n",
    "                - 'assigned_condition': the most frequent condition per cell,\n",
    "                - 'total_fb_counts': total FB reads across all conditions,\n",
    "                - 'condition_counts_rate': % of reads supporting assigned condition.\n",
    "    \"\"\"\n",
    "    \n",
    "    df = pd.read_csv(path)\n",
    "    \n",
    "    # extract predicted phase from 'pred' column (contains chunk_#_phase)\n",
    "    df['assigned_condition'] = df['pred'].str.extract(r'chunk_\\d+_(.*)')\n",
    "    df = df[['cell_barcode', 'assigned_condition']]\n",
    "    \n",
    "    # obtain counts for each phase for each cell barcode\n",
    "    df = df.groupby(['cell_barcode', 'assigned_condition']).value_counts().reset_index(name='count')\n",
    "    df = df.pivot_table(index='cell_barcode', columns='assigned_condition', values='count', aggfunc='sum', fill_value=0)\n",
    "    df.columns.name = None\n",
    "\n",
    "    # select the most frequent condition (the one with the most reads)\n",
    "    df['assigned_condition'] = df.idxmax(axis=1)\n",
    "\n",
    "    # compute read percentages for each possible condition\n",
    "    df['total_fb_counts'] = df.drop(columns='assigned_condition').sum(axis=1)\n",
    "    df['condition_counts_rate'] = df.apply(lambda row: row[row['assigned_condition']] / row['total_fb_counts'], axis=1)\n",
    "\n",
    "    print(f\"Final FB map shape: {df.shape}\")\n",
    "    \n",
    "    return df.rename(columns=column_rename_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ddec998-4ac9-4457-b9f7-7b31d551d729",
   "metadata": {},
   "source": [
    "# User Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dfc8f8c5-5c5c-4261-b547-a8393dbbdbcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_folders = {\n",
    "    'hybrid': \"/scratch/indikar_root/indikar1/shared_data/hyb_epi2me/hybrid/hybrid.gene_raw_feature_bc_matrix/\",\n",
    "    'control': \"/nfs/turbo/umms-indikar/shared/projects/HSC/pipeline_outputs/cc_fibroblast_full/scfib/scfib.gene_raw_feature_bc_matrix/\"\n",
    "}\n",
    "\n",
    "# Choose which datasets to load\n",
    "load_datasets = {\n",
    "    'hybrid': True,\n",
    "    'control': True\n",
    "}\n",
    "\n",
    "# Choose whether to include feature barcode maps\n",
    "use_feature_data = {\n",
    "    'hybrid': True,\n",
    "    'control': True\n",
    "}\n",
    "\n",
    "# Feature barcode map config\n",
    "feature_paths = {\n",
    "    'hybrid': \"/scratch/indikar_root/indikar1/shared_data/library_2_tsb_strict/feature_barcodes/fbc_map.csv\",\n",
    "    'control': \"/scratch/indikar_root/indikar1/shared_data/sc_fib_tsb_rerun/feature_barcodes/fbc_map.csv\"\n",
    "}\n",
    "\n",
    "column_rename_maps = {\n",
    "    \"hybrid\": {\"MYOD\": \"MYOD-fb_counts\", \"PRRX1\": \"PRRX1-fb_counts\", \"PRRX1_MYOD\": \"PRRX1_MYOD-fb_counts\"},\n",
    "    \"control\": {\"S\": \"S-fb_counts\", \"G2M\": \"G2M-fb_counts\", \"G1\": \"G1-fb_counts\"},\n",
    "}\n",
    "\n",
    "# Gene annotation file\n",
    "gtf_path = \"/nfs/turbo/umms-indikar/shared/projects/reference_genome/prebuilt/refdata-gex-GRCh38-2024-A/genes/genes.gtf\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "343e727d-1410-4eec-b16e-efb3ab268bf6",
   "metadata": {},
   "source": [
    "# Load in Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "28169c4f-848e-487a-ad73-97c928de0d2a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final FB map shape: (11639, 6)\n",
      "Final FB map shape: (10715, 6)\n",
      "\n",
      "Loading dataset: hybrid\n",
      "Feature data merged into 'hybrid' — cells per condition:\n",
      "assigned_condition\n",
      "PRRX1_MYOD    4174\n",
      "PRRX1         3955\n",
      "MYOD          2765\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Total cells in 'hybrid': 10895\n",
      "\n",
      "Loading dataset: control\n",
      "Feature data merged into 'control' — cells per condition:\n",
      "assigned_condition\n",
      "G1     5845\n",
      "G2M    1528\n",
      "S      1502\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Total cells in 'control': 8963\n"
     ]
    }
   ],
   "source": [
    "### Load feature barcode data\n",
    "feature_data = {}\n",
    "for key in main_folders:\n",
    "    if use_feature_data.get(key, False):\n",
    "        feature_data[key] = feature_barcode_map(feature_paths[key], column_rename_maps[key])\n",
    "\n",
    "### Load gene expression datasets (EPI2ME raw gene counts)\n",
    "datasets = {}\n",
    "for key, path in main_folders.items():\n",
    "    if load_datasets.get(key, False):\n",
    "        print(f\"\\nLoading dataset: {key}\")\n",
    "        mat = sc.read_10x_mtx(f\"{path}\")\n",
    "        mat.obs.index.name = \"cell_barcode\"\n",
    "        mat.var.index.name = \"gene_name\"\n",
    "        mat.obs_names = mat.obs_names.str.split(\"-\").str[0]\n",
    "\n",
    "        # Merge in FB data if selected\n",
    "        if use_feature_data.get(key, False) and key in feature_data:\n",
    "            mat.obs = mat.obs.merge(feature_data[key], left_index=True, right_index=True, how='left')\n",
    "            print(f\"Feature data merged into '{key}' — cells per condition:\")\n",
    "            print(mat.obs['assigned_condition'].value_counts())\n",
    "            print(f\"\\nTotal cells in '{key}': {mat.n_obs}\")\n",
    "        else:\n",
    "            print(f\"\\nNo feature data merged into '{key}'.\")\n",
    "            print(f\"\\nTotal cells in '{key}': {mat.n_obs}\")\n",
    "\n",
    "        mat.obs_names = [f\"{cb}_{key}\" for cb in mat.obs_names] # Make cell barcodes unique for merging\n",
    "        \n",
    "        datasets[key] = mat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1a888b1-2372-472d-8f59-ca9e3f333587",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Combine Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "228977c3-1c67-491d-bfde-b256b6163595",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Combined 2 datasets.\n",
      "Total cells in final AnnData: 19858\n",
      "Final combined AnnData shape: (19858, 28702)\n"
     ]
    }
   ],
   "source": [
    "### Combine the datasets\n",
    "if len(datasets) == 1:\n",
    "    adata = list(datasets.values())[0]\n",
    "    print(\"\\nOnly one dataset loaded.\")\n",
    "else:\n",
    "    adata = an.concat(\n",
    "        list(datasets.values()),\n",
    "        join=\"outer\",\n",
    "        label=\"dataset\",\n",
    "        keys=[k.capitalize() for k in datasets.keys()],\n",
    "        merge=\"same\",\n",
    "    )\n",
    "    print(f\"\\nCombined {len(datasets)} datasets.\")\n",
    "\n",
    "\n",
    "adata.obs['total_reads'] = adata.X.sum(axis=1) # number of reads per cell\n",
    "adata.obs['total_genes'] = (adata.X > 0).sum(axis=1) # number of genes detected per cell\n",
    "\n",
    "print(f\"Total cells in final AnnData: {adata.n_obs}\")\n",
    "print(f\"Final combined AnnData shape: {adata.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f48095-5e7b-4083-aeae-42c392d17975",
   "metadata": {},
   "source": [
    "# Filter and Merge Gene Annotation Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3826c9fb-9031-4349-9344-03dc39ad25e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting GTF processing...\n",
      "\n",
      "Raw GTF loaded. Shape: (3293161, 27)\n",
      "\n",
      "Filtered for 'gene' features: 38606 entries\n",
      "\n",
      "Matched with AnnData var names: 28723 entries\n",
      "\n",
      "Final gene annotation shape: (28702, 6)\n",
      "\n",
      "Merging transcript annotations (gdf) into adata.var...\n",
      "Gene annotations successfully merged into adata.var.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 19858 × 28702\n",
       "    obs: 'MYOD-fb_counts', 'PRRX1-fb_counts', 'PRRX1_MYOD-fb_counts', 'assigned_condition', 'total_fb_counts', 'condition_counts_rate', 'G1-fb_counts', 'G2M-fb_counts', 'S-fb_counts', 'dataset', 'total_reads', 'total_genes'\n",
       "    var: 'gene_id', 'gene_type', 'Chromosome', 'Start', 'End'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Starting GTF processing...\")\n",
    "\n",
    "gtf_raw = pr.read_gtf(gtf_path).as_df()\n",
    "print(f\"\\nRaw GTF loaded. Shape: {gtf_raw.shape}\")\n",
    "\n",
    "# Filter for 'gene' features\n",
    "gdf = gtf_raw[gtf_raw['Feature'] == 'gene']\n",
    "print(f\"\\nFiltered for 'gene' features: {len(gdf)} entries\")\n",
    "\n",
    "# Extract only genes that are present in adata\n",
    "gdf = gdf[gdf['gene_name'].isin(adata.var_names)]\n",
    "print(f\"\\nMatched with AnnData var names: {len(gdf)} entries\")\n",
    "\n",
    "# Select columns to include in gene annotation\n",
    "keep_columns = ['gene_id', 'gene_name', 'gene_type', 'Chromosome', 'Start', 'End']\n",
    "gdf = gdf[keep_columns].drop_duplicates(subset='gene_name').dropna(subset=['gene_name']).reset_index(drop=True)\n",
    "\n",
    "print(f\"\\nFinal gene annotation shape: {gdf.shape}\")\n",
    "\n",
    "# Merge with adata.var\n",
    "print(\"\\nMerging transcript annotations (gdf) into adata.var...\")\n",
    "adata.var = pd.concat([\n",
    "    adata.var,\n",
    "    gdf.set_index('gene_name')\n",
    "], axis=1)\n",
    "\n",
    "print(\"Gene annotations successfully merged into adata.var.\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f96330c7-d79c-44a3-bc66-15bd49e3eaae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MYOD-fb_counts</th>\n",
       "      <th>PRRX1-fb_counts</th>\n",
       "      <th>PRRX1_MYOD-fb_counts</th>\n",
       "      <th>assigned_condition</th>\n",
       "      <th>total_fb_counts</th>\n",
       "      <th>condition_counts_rate</th>\n",
       "      <th>G1-fb_counts</th>\n",
       "      <th>G2M-fb_counts</th>\n",
       "      <th>S-fb_counts</th>\n",
       "      <th>dataset</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>total_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGCAACTGC_hybrid</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>PRRX1_MYOD</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hybrid</td>\n",
       "      <td>18815.0</td>\n",
       "      <td>4890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGCTATGAT_hybrid</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>PRRX1_MYOD</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hybrid</td>\n",
       "      <td>1651.0</td>\n",
       "      <td>870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGTAGCCGT_hybrid</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>PRRX1_MYOD</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.954545</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hybrid</td>\n",
       "      <td>1928.0</td>\n",
       "      <td>1011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGTAGGGCA_hybrid</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MYOD</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hybrid</td>\n",
       "      <td>9679.0</td>\n",
       "      <td>3584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGTCTAGGC_hybrid</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>PRRX1_MYOD</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hybrid</td>\n",
       "      <td>28392.0</td>\n",
       "      <td>6359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         MYOD-fb_counts  PRRX1-fb_counts  \\\n",
       "AAACCAAAGCAACTGC_hybrid             0.0              0.0   \n",
       "AAACCAAAGCTATGAT_hybrid             1.0              1.0   \n",
       "AAACCAAAGTAGCCGT_hybrid             1.0              0.0   \n",
       "AAACCAAAGTAGGGCA_hybrid            14.0              1.0   \n",
       "AAACCAAAGTCTAGGC_hybrid             0.0              0.0   \n",
       "\n",
       "                         PRRX1_MYOD-fb_counts assigned_condition  \\\n",
       "AAACCAAAGCAACTGC_hybrid                  25.0         PRRX1_MYOD   \n",
       "AAACCAAAGCTATGAT_hybrid                   6.0         PRRX1_MYOD   \n",
       "AAACCAAAGTAGCCGT_hybrid                  21.0         PRRX1_MYOD   \n",
       "AAACCAAAGTAGGGCA_hybrid                   1.0               MYOD   \n",
       "AAACCAAAGTCTAGGC_hybrid                  25.0         PRRX1_MYOD   \n",
       "\n",
       "                         total_fb_counts  condition_counts_rate  G1-fb_counts  \\\n",
       "AAACCAAAGCAACTGC_hybrid             25.0               1.000000           NaN   \n",
       "AAACCAAAGCTATGAT_hybrid              8.0               0.750000           NaN   \n",
       "AAACCAAAGTAGCCGT_hybrid             22.0               0.954545           NaN   \n",
       "AAACCAAAGTAGGGCA_hybrid             16.0               0.875000           NaN   \n",
       "AAACCAAAGTCTAGGC_hybrid             25.0               1.000000           NaN   \n",
       "\n",
       "                         G2M-fb_counts  S-fb_counts dataset  total_reads  \\\n",
       "AAACCAAAGCAACTGC_hybrid            NaN          NaN  Hybrid      18815.0   \n",
       "AAACCAAAGCTATGAT_hybrid            NaN          NaN  Hybrid       1651.0   \n",
       "AAACCAAAGTAGCCGT_hybrid            NaN          NaN  Hybrid       1928.0   \n",
       "AAACCAAAGTAGGGCA_hybrid            NaN          NaN  Hybrid       9679.0   \n",
       "AAACCAAAGTCTAGGC_hybrid            NaN          NaN  Hybrid      28392.0   \n",
       "\n",
       "                         total_genes  \n",
       "AAACCAAAGCAACTGC_hybrid         4890  \n",
       "AAACCAAAGCTATGAT_hybrid          870  \n",
       "AAACCAAAGTAGCCGT_hybrid         1011  \n",
       "AAACCAAAGTAGGGCA_hybrid         3584  \n",
       "AAACCAAAGTCTAGGC_hybrid         6359  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e02ca92c-e5ef-4cae-b8a1-63162d38d4d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345177</td>\n",
       "      <td>58353492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58347717</td>\n",
       "      <td>58355455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr10</td>\n",
       "      <td>50799408</td>\n",
       "      <td>50885675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9067663</td>\n",
       "      <td>9116229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9065162</td>\n",
       "      <td>9068689</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id       gene_type Chromosome     Start       End\n",
       "gene_name                                                                \n",
       "A1BG       ENSG00000121410  protein_coding      chr19  58345177  58353492\n",
       "A1BG-AS1   ENSG00000268895          lncRNA      chr19  58347717  58355455\n",
       "A1CF       ENSG00000148584  protein_coding      chr10  50799408  50885675\n",
       "A2M        ENSG00000175899  protein_coding      chr12   9067663   9116229\n",
       "A2M-AS1    ENSG00000245105          lncRNA      chr12   9065162   9068689"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "76c55cc4-a76c-4bb0-953f-a1d9c4e3de56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 19858 × 28702\n",
       "    obs: 'MYOD-fb_counts', 'PRRX1-fb_counts', 'PRRX1_MYOD-fb_counts', 'assigned_condition', 'total_fb_counts', 'condition_counts_rate', 'G1-fb_counts', 'G2M-fb_counts', 'S-fb_counts', 'dataset', 'total_reads', 'total_genes'\n",
       "    var: 'gene_id', 'gene_type', 'Chromosome', 'Start', 'End'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adata.write(\"/scratch/indikar_root/indikar1/jrcwycy/HYB/anndata/new_raw_epi2me_merged.h5ad\")\n",
    "# adata"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
